Plugin system

6utt3rfly opened the pull request:
Staging area before releasing all plugins to main

6utt3rfly committed:
- Remove nested plugin folder (was originally thinking of tests next to source)
- Tests now restore original operation after running to avoid polluting other tests (since they modify static values versus instance values)
- filterProps test util now handles bool and null properly

6utt3rfly committed:
move CONDITIONAL_EXP to ternary plugin

6utt3rfly committed:
Replace `import {Jsep}` with `import jsep`

6utt3rfly committed:
Object plugin

6utt3rfly committed:
Update plugin system to support separate-files for each plugin with dependency injection
- Added plugin registration to JSEP. Plugins have a `name` and an `init` function that is called on registration with JSEP.
- Updated Object + Ternary plugins. Ternary plugin is registered by default
- Updated rollup to build each plugin as a separate output (all formats). Also delete dist folder before building.
- don't pollute the JSEP namespace from plugins. Use local constants instead.

6utt3rfly committed:
Update README.md with backticks

Co-authored-by: Lea Verou <lea@verou.me>

6utt3rfly committed:
Allow registering multiple plugins at once
(single object arg, single array of objects arg, multiple objects arg)

6utt3rfly committed:
Remove array arg support for plugin registration

6utt3rfly committed:
Rename plugins to prefix with `jsep`

6utt3rfly committed:
- rename `jsep.plugins.plugins` => `jsep.plugins.registered` and add to readme
- gobbleExpression and gobbleToken need to stop running hooks once a node is found, similar to how jsep checks when calling each successive gobble method now. This saves each hook from having to check for `env.node` (and potentially leading to future errors with 2 hooks parsing 2 sequential expressions as just the last node). Added unit tests.

6utt3rfly committed:
When testing, cleanup hooks before/after to avoid cross-interference between plugins and tests

6utt3rfly committed:
Spread operator plugin, ...a
- works in objects { ...a }, arrays [...a], function args fn(...a)
- NOTE: does not prevent `a ? ...b : ...c` or `...123`

6utt3rfly committed:
Arrow Function plugin
() => x
x => x
(a, b) => x

Treats '=>' as a binary operator, but replaces the binary expression with the correct node type afterward

6utt3rfly committed:
Assignment expression support plugin
a = 2
a *= 2

Treats each assignment operator as a binary operator, but replaces the binary expression with the 'AssignmentExpression' node type afterward

6utt3rfly committed:
Comment Plugin
- supports // and /* */ style
- throws an error on missing closing */

6utt3rfly committed:
'new' plugin (i.e. new Date(), new Class())

6utt3rfly committed:
Template Literal plugin

6utt3rfly committed:
Add Update assignment support (++/--)

6utt3rfly committed:
Merge pull request #145 from EricSmekens/plugin-obj-build

Update plugin system to use registration

6utt3rfly committed:
Updated package-lock with rollup-plugin-delete

6utt3rfly committed:
Merge pull request #144 from EricSmekens/plugin-object

Object plugin

6utt3rfly committed:
Merge pull request #149 from EricSmekens/plugin-comments

Comment Plugin

6utt3rfly committed:
Merge branch 'plugin-system' into plugin-new

6utt3rfly committed:
Merge pull request #150 from EricSmekens/plugin-new

'new' plugin

6utt3rfly committed:
Merge pull request #151 from EricSmekens/plugin-template-literal

Template Literal plugin

6utt3rfly committed:
Merge branch 'plugin-system' into plugin-spread

6utt3rfly committed:
Merge pull request #146 from EricSmekens/plugin-spread

Spread Operator Plugin

6utt3rfly committed:
Merge branch 'plugin-system' into plugin-assignment

6utt3rfly committed:
Merge branch 'plugin-system' into plugin-arrow

6utt3rfly commented:
@LeaVerou - I merged the approved PRs into this branch because I was thinking it would be better to ensure everything is working together and that we're happy with the API, documentation, etc. before merging all of it into master and releasing a new version. Buf if you would prefer to merge straight into main that's fine with me too 🤷🏻‍♀️

LeaVerou commented on README.md:
```diff
@@ -87,12 +87,66 @@ jsep.removeIdentifierChar('@');
+
+#### Writing Your Own Plugin:
+Plugins are objects with two properties: `name` and `init`.
+Here's a simple plugin example:
```
We should explain what this sample plugin does!

6utt3rfly replied:
Updated!

6utt3rfly committed:
Update the example plugin and add a description of what it does

EricSmekens commented:
Looks very good. I like the fact users can keep this simple, as I think most of jsep usages will be without plugins, but it is also very nice that extendability is possible and well documented now. Great!

6utt3rfly committed:
- Remove jsep.hooksAdd and just stick with hooks.add. This allows jsep to be reset easier (i.e. between unit tests)
- Rename resetJsepHooks => resetJsepDefaults (which currently includes jsepTernary plugin)
- Update typings

LeaVerou reviewed (approved):
LGTM from a quick look!

6utt3rfly commented:
Just waiting on #147 #148 and #153 before releasing (and bumping the version)?

6utt3rfly committed:
- Remove unnecessary `env.node` check (jsep.searchHook negates the need)

6utt3rfly committed:
Update version 0.4.0 => 1.0.0

6utt3rfly committed:
RegExp plugin:
/abc/
/abc/ig
/abc/.test(x)
/abc/ig['test'](x)

Adds a regex value type for 'Literal' expression types
- Note: modifies jsep to move the object .[( handling code into a function that the plugin can reuse (but code itself is functionally identical).

6utt3rfly committed:
Merge branch 'plugin-system' into plugin-arrow

6utt3rfly committed:
Merge branch 'plugin-system' into plugin-assignment

LeaVerou commented:
> Just waiting on #147 #148 and #153 before releasing (and bumping the version)?

Why do we need to have agreement on all individual plugins to merge the plugin system into jsep? The whole point of a plugin system is that it's modular, plugins can be added at any time, by us or others.

6utt3rfly commented:
> Why do we need to have agreement on all individual plugins to merge the plugin system into jsep? The whole point of a plugin system is that it's modular, plugins can be added at any time, by us or others.

I agree that we don't have to wait ... but the whole change to add the plugin system started from #123 and my original PR to add arrow-function support. So personally, I would like to have that included in the initial release... 🤷🏻‍♀️ . All of the plugins could realistically be moved into individual projects as well if that's preferred, but I think it's convenient, at the moment, to have it together to ensure everything works well together. I was also waiting on creating another PR that just adds a test file with all plugins combined, to ensure that all of the expressions and tests work when all plugins are registered together to ensure no errors with breaking other functionality with a loaded plugin.

6utt3rfly committed:
Add tests to confirm regex plugin does not interfere with binary operator `/` or `/=`

6utt3rfly committed:
Update to `resetJsepDefaults`

6utt3rfly committed:
Update to `resetJsepDefaults`
Add esprima comparison tests

LeaVerou commented:
I'm not used to the workflow of releasing a new version every time a PR is merged, but I'm not opposed to it either. 🤷🏼‍♀️ 
Ok, let's get consensus on the existing plugin PRs before merging. What's left to review?

6utt3rfly committed:
Merge pull request #155 from EricSmekens/plugin-regex

RegExp plugin:

6utt3rfly committed:
Merge pull request #147 from EricSmekens/plugin-arrow

Arrow Function plugin

6utt3rfly committed:
Fixes in build:
- rollup should use src/index.js so ternary plugin is included by default
- fix type defs

6utt3rfly committed:
Merge branch 'plugin-system' into plugin-assignment

# Conflicts:
#	README.md
#	rollup.config.js
#	test/tests.js

6utt3rfly committed:
Update assignment operator precedence to be near lowest priority
(see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence)

6utt3rfly commented:
> not used to the workflow of releasing a new version every time a PR is merged

I typically don't either ... but I thought for a significant change like this it might make more sense. Especially if we're updating the readme docs but haven't yet published to npm 🤷🏻‍♀️ 

At any rate, I think it's nearly ready. I just updated the .npmignore file after I noticed what was included (`npm pack` builds a tar file to test).

One issue I noticed, when trying to use the build in a typescript project, is that the plugins themselves don't have a type definition:
```
import * as jsep from 'jsep';
import * as jsepArrow from 'jsep/dist/cjs/jsepArrow.cjs.js';

TS7016: Could not find a declaration file for module
'jsep/dist/cjs/jsepArrow.cjs.js'. 'node_modules/jsep/dist/cjs/jsepArrow.cjs.js' implicitly has an 'any' type.   Try `npm i --save-dev @types/jsep` if it exists or add a new declaration (.d.ts) file containing `declare module 'jsep/dist/cjs/jsepArrow.cjs.js';`
```

I could fix it by adding types to the tsd.ts file:
```
declare module 'jsep/dist/jsepArrow' {
  import * as jsep from 'jsep';
  const name: string;
  const init: (this: typeof jsep) => void;
}

declare module 'jsep/dist/cjs/jsepArrow.cjs.js' {
  import * as jsep from 'jsep';
  const name: string;
  const init: (this: typeof jsep) => void;
}

declare module 'jsep/dist/jsepArrow' {
  import * as jsep from 'jsep';
  const name: string;
  const init: (this: typeof jsep) => void;
}

declare module 'jsep/dist/iife/jsepArrow.iife.js' {
  import * as jsep from 'jsep';
  const name: string;
  const init: (this: typeof jsep) => void;
}
```
Or do you have a better suggestion? (I'd hate to do nothing and force 'allow implicit any', or have to manually declare the module)

LeaVerou commented:
Can we have separate typings per module, included in its own directory? I'd like them to be as self-contained as possible.

6utt3rfly committed:
Merge pull request #148 from EricSmekens/plugin-assignment

assignment/update Plugin

6utt3rfly committed:
Add combined plugins tests to ensure cross-compatibility when using multiple plugins

6utt3rfly commented:
Maybe we should make the plugins more like a mono repo then? So that each plugin is individually published to npm (i.e. `@jsep/arrow-plugin`)? Might be nicer for versioning in the future if each has their own version? See #158

6utt3rfly committed:
Move to a mono-repo type structure for plugins. Each plugin is its own folder with its own package.json, rollup, etc. This keeps them more independent of jsep and allows for separate versioning

6utt3rfly committed:
Reduce boilerplate:
- remove .gitignore (root file works)
- remove rollup.config.js (and use a single file at the root)

Add plugin-specific types

Fix New plugin adding 'new' to a MemberExpression instead of the CallExpression

6utt3rfly committed:
Rename `plugins` folder to `packages` and update references

6utt3rfly committed:
Merge pull request #158 from EricSmekens/plugin-mono-repo

Plugin mono repo

6utt3rfly commented:
@LeaVerou and @EricSmekens - I think this should have all of the PRs and changes that were already individually reviewed and approved. Do you want one more review of the branch before merging all to main? With the separate packages per plugin, should the github actions be updated?

EricSmekens commented:
Approval on my side on merging this all to main.

Thank you for all your time and work on this! I think the actions should be updated, but I'm fine with that in a seperate pr.
(It currently uses npm publish, I guess for the multiple packages we need something different).

6utt3rfly commented:
> I guess for the multiple packages we need something different).

I think each plugin package could be manually published (`cd packages/arrow && npm publish`) in the meantime? :-)

6utt3rfly committed:
Fix cjs package.json type for package builds (module => commonjs)

6utt3rfly committed:
Type correction

6utt3rfly committed:
Fix hook types

LeaVerou commented:
Approval on my side too. We can always iterate with other PRs.

6utt3rfly committed:
Add `publishConfig.access: public` to help with first npm publish

6utt3rfly commented:
FYI - I started working on updating expression-eval to support the new node types and have been confirming that it can evaluate the nodes and use the updated jsep and packages. Merging! :-)

dereklieu commented:
This is exciting, I'm looking forward to using the `object` plugin. Do you have plans to include this work in an upcoming release?
